### PR TITLE
Add a mass sampler for the colossus halo mass function

### PR DESCRIPTION
**Pull request by sutieng**
## Checklist
- [x] Follow the [Contributor Guidelines](https://github.com/skypyproject/skypy/blob/master/CONTRIBUTING.md)
- [ ] Write unit tests
- [x] Write documentation strings
- [ ] Assign someone from your working team to review this pull request
- [ ] Assign someone from the infrastructure team to review this pull request

### CODE REVIEW

File: skypy/clusters/sampler.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        return a* (1+scipy.special.erf(((x)- b ) / c )  ) + d
+    
+    if z <= 0.1:  ##apply mass cut: remove halo with halo mass M500c<1e+13
+        fit_M = 1e+13
+    else:
+        p = func(z)
+        fit_M = 10**(p)
+    return fit_M  ##unit: Msun
+
+def mass_sampler(z, size, mdef = '500c', m_min = 1e+12, m_max = 1e+16, select = False, newcosmo_astropy = WMAP9, sigma8 = 0.8200, ns = 0.9608):
```
Comment (rrjbca) [inline skypy/clusters/sampler.py:46]:
```suggestion
def colossus_mass_sampler(redshift, model, mdef, m_min, m_max, cosmology, sigma8, ns, size=None):
```
- Rename the function for clarity
- Rename `z` to `redshift` for clarity
- `size=None` should have the same default value as for `np.random.uniform`
- Add `resolution=1000` as a default argument that the user can adjust rather than hardcoding
- No other arguments should have default values
- Make the function more general by taking the colossus `model` as an argument as well
- Remove the selection effect from this sampler. It is too specific (works for eRosita only) and we can calculate it in a separate step in the pipeline since it effectively just determines `m_min`
- Rename `newcosmo_astropy` to `cosmology` since the SkyPy Pipeline class recognises this name
- Update the docstrings to reflect these changes

File: skypy/clusters/sampler.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    cosmo = colossus.cosmology.cosmology.fromAstropy(newcosmo_astropy, sigma8 = sigma8, ns = ns, name = 'my_cosmo')
+    h0 = newcosmo_astropy.h
+    m_h0 = np.logspace(np.log10(m_min*h0), np.log10(m_max*h0), 1000) ##unit: Msun/h
+    dndm = mass_function.massFunction(m_h0, z, mdef = mdef, model = 'despali16',q_out = 'dndlnM',q_in='M')*h0**3/m_h0*h0  ##unit: 1/Msun/Mpc^3
+    m = m_h0/h0
+    if select==True:
+        dndm_se = dndm*(np.heaviside([m - selection(z)], 1.0))
+        dndm_se = np.reshape(dndm_se, (len(m)))
+    else:
+        dndm_se = dndm
```
Comment (rrjbca) [inline skypy/clusters/sampler.py:95]:
```suggestion
```
Remove the selection effect from this function we can do it separately in the pipeline

File: skypy/clusters/sampler.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    >>>from scipy import integrate
+
+    >>>sigma8 = 0.8200
+    >>>ns = 0.9608
+    >>>mass_sampler(z=0.1,size=100,select=True)
+    
+    """
+    cosmo = colossus.cosmology.cosmology.fromAstropy(newcosmo_astropy, sigma8 = sigma8, ns = ns, name = 'my_cosmo')
+    h0 = newcosmo_astropy.h
+    m_h0 = np.logspace(np.log10(m_min*h0), np.log10(m_max*h0), 1000) ##unit: Msun/h
```
Comment (rrjbca) [inline skypy/clusters/sampler.py:88]:
```suggestion
    m_h0 = np.logspace(np.log10(m_min*h0), np.log10(m_max*h0), resolution) ##unit: Msun/h
```

File: skypy/clusters/sampler.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        c = 1.52774801e+00
+        d = -1.85377292e+05
+        return a* (1+scipy.special.erf(((x)- b ) / c )  ) + d
+    
+    if z <= 0.1:  ##apply mass cut: remove halo with halo mass M500c<1e+13
+        fit_M = 1e+13
+    else:
+        p = func(z)
+        fit_M = 10**(p)
+    return fit_M  ##unit: Msun
```
Comment (rrjbca) [inline skypy/clusters/sampler.py:44]:
```suggestion
```
Remove this for now we will handle the selection function separately

File: skypy/clusters/sampler.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+    >>>sigma8 = 0.8200
+    >>>ns = 0.9608
+    >>>mass_sampler(z=0.1,size=100,select=True)
+    
+    """
+    cosmo = colossus.cosmology.cosmology.fromAstropy(newcosmo_astropy, sigma8 = sigma8, ns = ns, name = 'my_cosmo')
+    h0 = newcosmo_astropy.h
+    m_h0 = np.logspace(np.log10(m_min*h0), np.log10(m_max*h0), 1000) ##unit: Msun/h
+    dndm = mass_function.massFunction(m_h0, z, mdef = mdef, model = 'despali16',q_out = 'dndlnM',q_in='M')*h0**3/m_h0*h0  ##unit: 1/Msun/Mpc^3
```
Comment (rrjbca) [inline skypy/clusters/sampler.py:89]:
```suggestion
    dndm = mass_function.massFunction(m_h0, redshift, mdef=mdef, model=model, q_out='dndlnM', q_in='M') / m_h0
```
No need to keep track of the factors of h0 units here since you normalise the CDF in the next step

**Review by rrjbca — CHANGES_REQUESTED** [review]
Looks good so far. A few suggestions to generalise the function and make it more useful. I would also suggest:
- This file should be in the halos directory and renamed. I suggest `git mv skypy/clusters/sampler.py skypy/halos/_colossus.py`.
- Update the docstring and docs/halos/index.rst to ensure the new function appears in the documentation
- Write some unit tests. Perhaps look at the tests for `schechter_lf_magnitude` as that is quite a similar function

If you need help with any of this just ask

File: skypy/halos/sampler.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    >>>from colossus.cosmology import cosmology
+    >>>from colossus.lss import mass_function
+    >>>from scipy import integrate
+
+    >>>sigma8 = 0.8200
+    >>>ns = 0.9608
+    >>>colossus_mass_sampler(redshift=0.1, model='despali16', mdef= '500c', m_min = 1e+12, m_max = 1e+16, cosmology=WMAP9, sigma8 = 0.8200, ns = 0.9608, size=100)
+    
+    """
+    cosmo = colossus.cosmology.cosmology.fromAstropy(cosmology, sigma8 = sigma8, ns = ns, name = 'my_cosmo')
```
Comment (Lucia-Fonseca) [inline skypy/halos/sampler.py:50]: Is the name `my_cosmo` necessary?
Is `cosmo` necessary?
```suggestion
```

File: skypy/halos/sampler.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    h0 = cosmo.h
```
Comment (Lucia-Fonseca) [inline skypy/halos/sampler.py:51]:
```suggestion
    h0 = cosmology.h
```

File: skypy/halos/sampler.py [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    m_h0 = np.logspace(np.log10(m_min*h0), np.log10(m_max*h0), resolution) ##unit: Msun/h
```
Comment (Lucia-Fonseca) [inline skypy/halos/sampler.py:52]:
```suggestion
    m_h0 = np.logspace(np.log10(m_min*h0), np.log10(m_max*h0), resolution) # unit: Msun/h
```

**Review by Lucia-Fonseca — CHANGES_REQUESTED** [review]
- Agreed with @rrjbca 's comments.
- Please, check the flake8 style error messages.
- Can you add the equation number in Dispali+16 as a reference?

File: skypy/halos/sampler.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
+'''Halo mass sampler.
+This code samples cluster halos from colossus mass function .
+
+'''
+import numpy as np
+import colossus as colossus
+from colossus.lss import mass_function
+from scipy import integrate
+
+def colossus_mass_sampler(redshift, model, mdef, m_min, m_max, cosmology, sigma8, ns, size=None, resolution=1000):
```
[the remark this answers is not in the document]
Comment (rrjbca) [reply skypy/halos/sampler.py:10]: No, this function is supposed to be able to sample from any mass function implemented in colossus not just despali16. Therefore the more general name `colossus_mass_sampler` makes sense and `model` is required as an argument. Also `mdef` is the variable name used by colossus and so will be more familiar to users.

File: skypy/halos/sampler.py [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+    """
+        This function generate a sample of cluster with halo mass from Despali+16 mass function.
+       
+    Parameters
+    -----------
+    redshift: float
```
[the remark this answers is not in the document]
Comment (rrjbca) [reply skypy/halos/sampler.py:16]: No, in colossus `massFunction` can only take a scalar redshift, see [here](https://bdiemer.bitbucket.io/colossus/lss_mass_function.html#lss.mass_function.massFunction)

File: skypy/halos/sampler.py [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+     Redshift
+    model: string
+     Mass function model which is available in colossus
+    mdef: str
```
[the remark this answers is not in the document]
Comment (rrjbca) [reply skypy/halos/sampler.py:20]: No, this is the variable name used in colossus for consistency

File: skypy/halos/sampler.py [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+def colossus_mass_sampler(redshift, model, mdef, m_min, m_max, cosmology, sigma8, ns, size=None, resolution=1000):
+    """
+        This function generate a sample of cluster with halo mass from Despali+16 mass function.
+       
+    Parameters
+    -----------
+    redshift: float
+     Redshift
+    model: string
+     Mass function model which is available in colossus
```
[the remark this answers is not in the document]
Comment (rrjbca) [reply skypy/halos/sampler.py:19]: Yes we need `model`, we want to be able to sample from any halo mass function implemented in colossus not just despali16

File: skypy/halos/sampler.py (hunk 6 above)
Comment (rrjbca) [reply skypy/halos/sampler.py:50]: This line of code is essential; we have to register the cosmology we are using with colossus

File: skypy/halos/sampler.py [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+This code samples cluster halos from colossus mass function .
+
+'''
+import numpy as np
+import colossus as colossus
+from colossus.lss import mass_function
+from scipy import integrate
+
+def colossus_mass_sampler(redshift, model, mdef, m_min, m_max, cosmology, sigma8, ns, size=None, resolution=1000):
+    """
```
[the remark this answers is not in the document]
Comment (rrjbca) [reply skypy/halos/sampler.py:11]:
```
    """Colossus halo mass sampler.
```

File: skypy/halos/sampler.py [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+        This function generate a sample of cluster with halo mass from Despali+16 mass function.
```
[the remark this answers is not in the document]
Comment (rrjbca) [reply skypy/halos/sampler.py:12]: We shouldn't be referencing the the Despali+16 paper; we should use intersphinx to link to the documentation for [`massFunction`](https://bdiemer.bitbucket.io/colossus/lss_mass_function.html#lss.mass_function.massFunction) in colossus

File: skypy/halos/sampler.py [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+       
+    Parameters
+    -----------
+    redshift: float
+     Redshift
+    model: string
+     Mass function model which is available in colossus
+    mdef: str
+     halo mass definition, any spherical overdensity. For example: '200m', 200c', '500c', '2500c'
+         ***If select=True, then mdef has to be '500c'.
```
[the remark this answers is not in the document]
Comment (rrjbca) [reply skypy/halos/sampler.py:22]: This list doesn't cover all of the values for `mdef` accepted by colossus e.g. `fof`, `vir`...

[the remark this answers is not in the document]
Comment (rrjbca) [reply skypy/halos/sampler.py:22]: Also remove the reference to `select` as this has now been removed

File: skypy/halos/sampler.py [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+    m_min, m_max: float
+     Lower and upper bounds for cluster mass (Unit: Msun).
+    cosmology: astropy.cosmology.Cosmology
+    sigma8: float
+    ns: float
+    size: int
+     Number of clusters we need.
```
[the remark this answers is not in the document]
Comment (rrjbca) [reply skypy/halos/sampler.py:29]: I prefer how the default behaviour is described in e.g. `numpy.random.uniform`: "If size is None (default), a single value is returned."

File: skypy/halos/sampler.py [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+    redshift: float
+     Redshift
+    model: string
+     Mass function model which is available in colossus
+    mdef: str
+     halo mass definition, any spherical overdensity. For example: '200m', 200c', '500c', '2500c'
+         ***If select=True, then mdef has to be '500c'.
+    m_min, m_max: float
+     Lower and upper bounds for cluster mass (Unit: Msun).
+    cosmology: astropy.cosmology.Cosmology
```
[the remark this answers is not in the document]
Comment (rrjbca) [reply skypy/halos/sampler.py:25]: I would just say "Astropy cosmology object"; they will always be valid since they all require H0 and Om0 by definition, but Ob0 is not strictly necessary. Also `omega_matter` and `omega_matter` don't match the variable names used by astropy so could be confusing.

File: skypy/halos/sampler.py [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+    >>>from astropy.cosmology import WMAP9
+    >>>from colossus.cosmology import cosmology
+    >>>from colossus.lss import mass_function
+    >>>from scipy import integrate
+
+    >>>sigma8 = 0.8200
+    >>>ns = 0.9608
+    >>>colossus_mass_sampler(redshift=0.1, model='despali16', mdef= '500c', m_min = 1e+12, m_max = 1e+16, cosmology=WMAP9, sigma8 = 0.8200, ns = 0.9608, size=100)
+    
+    """
```
[the remark this answers is not in the document]
Comment (rrjbca) [reply skypy/halos/sampler.py:49]: No, this function is not specific to the Despali model any more

**Review by rrjbca — COMMENTED** [review]
Would it be reasonable to refer to the objects sampled by this function as "halos" rather than "clusters" throughout the docstrings here? i.e. are "clusters" simply "high-mass halos" in this context?

**Comment by rrjbca** [comment]
@sutieng could you please add a "Description" for this pull request? We also need to write some unit tests and integrate the function docstring into the documentation, are you able to do these or would you like one of us to help?

File: skypy/halos/sampler.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Lucia-Fonseca) [reply skypy/halos/sampler.py:10]: Agreed. Please ignore my previous suggestion.

File: skypy/halos/sampler.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (Lucia-Fonseca) [reply skypy/halos/sampler.py:11]: Agreed. Please ignore my previous suggestion.

File: skypy/halos/sampler.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (Lucia-Fonseca) [reply skypy/halos/sampler.py:12]: Agreed. Sorry for my confusion. Please ignore my previous suggestion.

File: skypy/halos/sampler.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (Lucia-Fonseca) [reply skypy/halos/sampler.py:16]: Agreed. Please ignore.

File: skypy/halos/sampler.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (Lucia-Fonseca) [reply skypy/halos/sampler.py:19]: Agreed.

File: skypy/halos/sampler.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (Lucia-Fonseca) [reply skypy/halos/sampler.py:20]: Agreed.

File: skypy/halos/sampler.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (Lucia-Fonseca) [reply skypy/halos/sampler.py:49]: Agreed. Should we have `Colossus` as reference?

File: skypy/halos/sampler.py [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+    >>>import astropy as astropy
+    >>>from astropy.cosmology import WMAP9
+    >>>from colossus.cosmology import cosmology
+    >>>from colossus.lss import mass_function
+    >>>from scipy import integrate
+
+    >>>sigma8 = 0.8200
+    >>>ns = 0.9608
+    >>>colossus_mass_sampler(redshift=0.1, model='despali16', mdef= '500c', m_min = 1e+12, m_max = 1e+16, cosmology=WMAP9, sigma8 = 0.8200, ns = 0.9608, size=100)
+
```
[the remark this answers is not in the document]
Comment (sutieng) [reply skypy/halos/sampler.py:49]: Thanks. I will remove this part.

**Comment by sutieng** [comment]
> Would it be reasonable to refer to the objects sampled by this function as "halos" rather than "clusters" throughout the docstrings here? i.e. are "clusters" simply "high-mass halos" in this context?
Yes, I agree. Since here we are only considering DM component, so cluster can be referred to high mass halos. I will make changes of the docstrings.

File: skypy/halos/sampler.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (sutieng) [reply skypy/halos/sampler.py:50]: We can have the code paper Diemer et al. 2018 as reference.

File: skypy/halos/sampler.py [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+    Parameters
+    -----------
+    redshift: float
+     Redshift
+    model: string
+     Mass function model which is available in colossus
+    mdef: str
+     halo mass definition, any spherical overdensity. For example: '200m', 200c', '500c', '2500c'
+         ***If select=True, then mdef has to be '500c'.
+    m_min, m_max: float
```
[the remark this answers is not in the document]
Comment (sutieng) [reply skypy/halos/sampler.py:24]: np.interp only accept 1-D sequence, so m_h0 has to be a 1-D array. Input for m_min and m_max need be a single value for now.

**Comment by sutieng** [comment]
> @sutieng could you please add a "Description" for this pull request? We also need to write some unit tests and integrate the function docstring into the documentation, are you able to do these or would you like one of us to help?
I don't have experience on writing unit tests. It would be great if you could help and let me know what items/ tests are needed to be included in unit tests.

File: setup.cfg [hunk 21]
```diff
@@ -34,6 +34,7 @@ test =
     speclite>=0.11
 all =
     speclite>=0.11
+    colossus
 docs =
     sphinx-astropy
     matplotlib
```
Comment (rrjbca) [inline setup.cfg:40]:
```suggestion
    matplotlib
    colossus
```
This is needed so ReadTheDocs will install colossus before trying to build the documentation

File: skypy/halos/sampler.py [hunk 22]
```diff
@@ -0,0 +1,75 @@
...
+    :nosignatures:
+    :toctree: ../api/
+
+    colossus_mass_sampler
+
+"""
+
+import numpy as np
+import colossus as colossus
+from colossus.lss import mass_function
```
Comment (rrjbca) [inline skypy/halos/sampler.py:18]:
```suggestion
```
If colossus is not installed this could raise an error. We will import them later instead...

File: skypy/halos/sampler.py [hunk 23]
```diff
@@ -0,0 +1,75 @@
...
+from scipy import integrate
+
+__all__ = [
+        'colossus_mass_sampler',
+           ]
+
+
```
Comment (rrjbca) [inline skypy/halos/sampler.py:25]:
```suggestion
try:
   import colossus
except ImportError:
   HAS_COLOSSUS = False
else:
   HAS_COLOSSUS = True


```
Defines a variable `HAS_COLOSSUS` which is True/False depending on if colossus is installed

File: skypy/halos/sampler.py [hunk 24]
```diff
@@ -0,0 +1,75 @@
...
+    Returns
+    --------
+    sample : (size,) array_like
+        Samples drawn from the mass function, in units of solar masses.
+
+        References
+    -----------
+    .. [1] Diemer et al. (2018) doi 10.3847/1538-4365/aaee8c
+    """
+    cosmo = colossus.cosmology.cosmology.fromAstropy(cosmology, sigma8=sigma8,
```
Comment (rrjbca) [inline skypy/halos/sampler.py:64]:
```suggestion
    from colossus.lss import mass_function
    cosmo = colossus.cosmology.cosmology.fromAstropy(cosmology, sigma8=sigma8,
```
Move import inside the function so that if somebody tries to call the function without installing colossus an ImportError will be raised

File: skypy/halos/tests/test_sampler.py [hunk 25]
```diff
@@ -0,0 +1,32 @@
+import numpy as np
+import colossus as colossus
+from colossus.lss import mass_function
```
Comment (rrjbca) [inline skypy/halos/tests/test_sampler.py:3]:
```suggestion
```
Again we will import this later to avoid any errors if colossus is not installed

File: skypy/halos/tests/test_sampler.py [hunk 26]
```diff
@@ -0,0 +1,32 @@
...
+from scipy import integrate
+from scipy.stats import kstest
+import pytest
+
```
Comment (rrjbca) [inline skypy/halos/tests/test_sampler.py:7]:
```suggestion
from skypy.halos.sampler import HAS_COLOSSUS

```
We want to import the `HAS_COLOSSUS` variable so we know if colossus is installed or not

File: skypy/halos/tests/test_sampler.py [hunk 27]
```diff
@@ -0,0 +1,32 @@
...
+
+@pytest.mark.flaky
```
Comment (rrjbca) [inline skypy/halos/tests/test_sampler.py:9]:
```suggestion
@pytest.mark.skipif(not HAS_COLOSSUS, reason='test requires colossus')
@pytest.mark.flaky
```
Only run the test if colossus is installed (i.e. skip if `HAS_COLOSSUS==False)

File: skypy/halos/tests/test_sampler.py [hunk 28]
```diff
@@ -0,0 +1,32 @@
...
+def test_colossus_mass_sampler():
+    from astropy.cosmology import WMAP9
```
Comment (rrjbca) [inline skypy/halos/tests/test_sampler.py:11]:
```suggestion
    from astropy.cosmology import WMAP9
    import colossus as colossus
    from colossus.lss import mass_function
```
import colossus inside the test so that no ImportErrors are raised if it is not installed

File: skypy/halos/tests/test_sampler.py [hunk 29]
```diff
@@ -0,0 +1,32 @@
...
+    from skypy.halos.sampler import colossus_mass_sampler
+    m_min = 1e+12
+    sample = colossus_mass_sampler(redshift=0.1, model='despali16',
+                                   mdef='500c', m_min=m_min, m_max=1e+16,
+                                   cosmology=WMAP9, sigma8=0.8200, ns=0.9608,
+                                   size=100, resolution=1000)
+    assert np.all(sample >= m_min)
```
Comment (rrjbca) [inline skypy/halos/tests/test_sampler.py:18]:
```suggestion
    m_min, m_max, size = 1e+12, 1e+16, 100
    sample = colossus_mass_sampler(redshift=0.1, model='despali16',
                                   mdef='500c', m_min=m_min, m_max=m_max,
                                   cosmology=WMAP9, sigma8=0.8200, ns=0.9608,
                                   size=size, resolution=1000)
    assert np.all(sample >= m_min)
    assert np.all(sample <= m_max)
    assert np.shape(sample) == (size,)
```
Test that samples are not greater than m_max and that the correct number of samples are returned

**Review by rrjbca — CHANGES_REQUESTED** [review]
Thanks for writing the tests and documentation they look great. Just a few suggestions:
- Move imports and skip unit test to avoid an ImportError if colossus is not installed
- Make sure to install colossus on ReadTheDocs
- Also test the maximum value and the shape of the samples

File: skypy/halos/sampler.py (hunk 23 above)
Comment (sutieng) [reply skypy/halos/sampler.py:23]: Thanks for this suggestion!

File: skypy/halos/tests/test_sampler.py (hunk 29 above)
Comment (sutieng) [reply skypy/halos/tests/test_sampler.py:18]: Thanks for these suggested tests. Will add them to the code!

File: skypy/halos/sampler.py [hunk 30]
```diff
@@ -0,0 +1,81 @@
...
+    HAS_COLOSSUS = False
+else:
+    HAS_COLOSSUS = True
+
+
+def colossus_mass_sampler(redshift, model, mdef, m_min, m_max,
+                          cosmology, sigma8, ns, size=None, resolution=1000):
+    """Colossus halo mass sampler.
+        This function generate a sample of halos from a mass function which
+        is available in colossus.
```
Comment (rrjbca) [inline skypy/halos/sampler.py:35]:
```suggestion
    """Colossus halo mass sampler.
    
    This function generate a sample of halos from a mass function which
    is available in COLOSSUS [1]_.
```

File: skypy/halos/sampler.py [hunk 31]
```diff
@@ -0,0 +1,81 @@
...
+    Parameters
+    -----------
+    redshift : float
+        The redshift values at which to sample the halo mass.
+    model : string
+        Mass function model which is available in colossus.
+    mdef : str
+        Halo mass definition for spherical overdensities.
+        Choose from :math:`<int>c`, :math:`<int>m`, :math:`vir`, :math:`fof`.
+        For example: '200m', 200c', '500c', 'fof', 'vir'.
```
Comment (rrjbca) [inline skypy/halos/sampler.py:46]:
```suggestion
        Halo mass definition for spherical overdensities used by colossus.
```

File: skypy/halos/sampler.py [hunk 32]
```diff
@@ -0,0 +1,81 @@
...
+    m_min, m_max : float
+        Lower and upper bounds for halo mass in units of Solar mass, :math:`Msun`.
```
Comment (rrjbca) [inline skypy/halos/sampler.py:48]:
```suggestion
        Lower and upper bounds for halo mass in units of Solar mass, Msun.
```

File: skypy/halos/sampler.py [hunk 33]
```diff
@@ -0,0 +1,81 @@
...
+    cosmology : astropy.cosmology.Cosmology
+        Astropy cosmology object
+    sigma8 : float
+        Cosmology parameter, amplitude of the (linear) power spectrum on the
+        scale of :math:`8 h-1 Mpc`.
```
Comment (rrjbca) [inline skypy/halos/sampler.py:53]:
```suggestion
        scale of 8 Mpc/h.
```

File: skypy/halos/sampler.py [hunk 34]
```diff
@@ -0,0 +1,81 @@
...
+    ns : float
+        Cosmology parameter, spectral index of scalar perturbation power spectrum.
+    size : int, optional
+        Number of halos to sample. If size is None (default), a single value is returned.
+    resolution : int, optional
+        Resolution of the inverse transform sampling spline. Default is 1000.
+    Returns
```
Comment (rrjbca) [inline skypy/halos/sampler.py:60]:
```suggestion
        Resolution of the inverse transform sampling spline. Default is 1000.

    Returns
```

File: skypy/halos/sampler.py [hunk 35]
```diff
@@ -0,0 +1,81 @@
...
+    --------
+    sample : (size,) array_like
+        Samples drawn from the mass function, in units of solar masses.
+
+        References
+    -----------
+    .. [1] Diemer et al. (2018) doi 10.3847/1538-4365/aaee8c
+    """
```
Comment (rrjbca) [inline skypy/halos/sampler.py:68]:
```suggestion
    References
    ----------
    .. [1] Diemer B., 2018, ApJS, 239, 35

    """
```

**Review by rrjbca — CHANGES_REQUESTED** [review]
Excellent, the tests are running and the documentation is building now. Just a couple of small final changes:
- If you look at the documentation [here](https://skypy--423.org.readthedocs.build/en/423/api/skypy.halos.sampler.colossus_mass_sampler.html#skypy.halos.sampler.colossus_mass_sampler) you can see some parts have not rendered correctly, mostly due to whitespace and units. See suggestions for how to improve this.
- Since this is your first contribution could you add your details to [.zenodo.json](https://github.com/skypyproject/skypy/blob/main/.zenodo.json) as a creator

File: skypy/halos/_colossus.py [hunk 36]
```diff
@@ -0,0 +1,74 @@
...
+    resolution : int, optional
+        Resolution of the inverse transform sampling spline. Default is 1000.
+
+    Returns
+    --------
+    sample : (size,) array_like
+        Samples drawn from the mass function, in units of solar masses.
+
+        References
+    -----------
```
Comment (rrjbca) [inline skypy/halos/_colossus.py:58]:
```suggestion
    References
    ----------
```

File: skypy/halos/_colossus.py [hunk 37]
```diff
@@ -0,0 +1,74 @@
...
+        scale of 8 Mpc/h.
+    ns : float
+        Cosmology parameter, spectral index of scalar perturbation power spectrum.
+    size : int, optional
+        Number of halos to sample. If size is None (default), a single value is returned.
+    resolution : int, optional
+        Resolution of the inverse transform sampling spline. Default is 1000.
+
+    Returns
+    --------
```
Comment (rrjbca) [inline skypy/halos/_colossus.py:53]:
```suggestion
    Returns
    -------
```

File: skypy/halos/_colossus.py [hunk 38]
```diff
@@ -0,0 +1,74 @@
...
+
+def colossus_mass_sampler(redshift, model, mdef, m_min, m_max,
+                          cosmology, sigma8, ns, size=None, resolution=1000):
+    """Colossus halo mass sampler.
+
+    This function generate a sample of halos from a mass function which
+    is available in COLOSSUS [1]_.
+
+    Parameters
+    -----------
```
Comment (rrjbca) [inline skypy/halos/_colossus.py:31]:
```suggestion
    Parameters
    ----------
```

**Review by rrjbca — APPROVED** [review]
Looks good to me, thank you for working on this @sutieng

**Review by Lucia-Fonseca — APPROVED** [review]
This works perfectly.
Great job @sutieng!
